pkg/source/service.go

```diff
@@ -0,0 +1,108 @@
+package source
+
+import (
+	"context"
+	"time"
+
+	"google.golang.org/protobuf/types/known/emptypb"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	sourcepb "github.com/numaproj/numaflow-go/pkg/apis/proto/source/v1"
+	"github.com/numaproj/numaflow-go/pkg/source/model"
+)
+
+// Service implements the proto gen server interface
+type Service struct {
+	sourcepb.UnimplementedSourceServer
+	Source Source
+}
+
+// IsReady returns true to indicate the gRPC connection is ready.
+func (fs *Service) IsReady(context.Context, *emptypb.Empty) (*sourcepb.ReadyResponse, error) {
+	return &sourcepb.ReadyResponse{Ready: true}, nil
+}
+
+// PendingFn returns the number of pending messages.
+func (fs *Service) PendingFn(ctx context.Context, _ *emptypb.Empty) (*sourcepb.PendingResponse, error) {
+	return &sourcepb.PendingResponse{Result: &sourcepb.PendingResponse_Result{
+		Count: fs.Source.Pending(ctx),
+	}}, nil
+}
+
+// readRequest implements the ReadRequest interface and is used in the read handler.
+type readRequest struct {
+	count   uint64
+	timeout time.Duration
+}
+
+func (r *readRequest) TimeOut() time.Duration {
+	return r.timeout
+}
+
+func (r *readRequest) Count() uint64 {
+	return r.count
+}
+
+// ReadFn reads the data from the source.
+func (fs *Service) ReadFn(d *sourcepb.ReadRequest, stream sourcepb.Source_ReadFnServer) error {
+	request := readRequest{
+		count:   d.Request.GetNumRecords(),
+		timeout: time.Duration(d.Request.GetTimeoutInMs()) * time.Millisecond,
+	}
+	ctx := stream.Context()
+	messageCh := make(chan model.Message)
+
+	// Start the read in a goroutine
+	go func() {
+		defer close(messageCh)
```

Comment:
Use the pattern in https://github.com/numaproj/numaflow-go/blob/main/pkg/function/service.go#L153-L167, this will solve the problem of the developer getting the channel closed in the `Read()` implementation.

Reply:
Context: We don't expect the `Read()` implementation to close the channel, but it is possible ppl do that, that that will make line 57 panic.

Reply:
We shouldn't let users close the channel and if they do that, let it panic. Ownership will get very tricky otherwise. I did meet with Keran, and we both agreed that we should have strict ownership principles. 

We have added a comment saying, please do not close :)